### ISSUE TITLE
org.gradle.internal.classpath.Instrumented not found when using System.getProperties() in a Gradle Plugin

### ISSUE THREAD

**Issue by rachelcarmena** [gradle/gradle#14727]
### Expected Behavior

No errors when building a project that uses a Kotlin Compiler Plugin during compilation.

That Kotlin Compiler Plugin is provided to the compilation via a custom Gradle Plugin.

It has been working since Gradle `5.0`until Gradle `6.5.1` (included).

### Current Behavior

```
$./gradlew clean build

> Task :compileKotlin FAILED
e: java.lang.NoClassDefFoundError: org/gradle/internal/classpath/Instrumented
  at arrow.meta.internal.registry.InternalRegistry$DefaultImpls.registerMetaComponents(InternalRegistry.kt:150)
```

`InternalRegistry.kt` comes from the Kotlin Compiler Plugin and the content of that line is:

```
println("System.properties are: " + System.getProperties().map {
```

### Context

It works until Gradle `6.5.1` (included). 

It fails with Gradle `6.6`, Gradle `6.6.1` or Gradle `6.7-rc-3`.

It impacts on environments with latest Gradle versions that try the examples of the Kotlin Compiler Plugin.

The error has been reported by a user in Kotlin Lang Slack - `#arrow-meta` channel.

We didn't realize about it because the example is using Gradle `5.6.4` and the first Gradle version was `5.0`.

### Steps to Reproduce 

With Gradle `6.6`, Gradle `6.6.1` or Gradle `6.7-rc-3`, run:

```
git clone https://github.com/arrow-kt/arrow-meta-examples.git
cd use-arrow-meta-compiler-plugin
gradle clean build
```

### Your Environment

Build scan URL: https://scans.gradle.com/s/doy7eoh223nlc

**Comment by rachelcarmena** [comment]
The error comes from the custom Gradle Plugin that provides the Kotlin Compiler Plugin.

No errors when providing the Kotlin Compiler Plugin manually via `compileKotlin.kotlinOptions.freeCompilerArgs` without the custom Gradle Plugin.

**Comment by rachelcarmena** [comment]
I've updated the title:

> org.gradle.internal.classpath.Instrumented not found when using System.getProperties() in a Gradle Plugin

**Comment by ljacomet** [comment]
Hey @eskatos, could you take a look? That seems linked to the configuration cache work.

**Comment by rachelcarmena** [comment]
Thank you so much @ljacomet :raised_hands:

**Comment by SheliakLyr** [comment]
Another example: https://github.com/trustin/sphinx-gradle-plugin/issues/19

**Comment by codingDriver** [comment]
We have the same problem. grade Plugin in groovy invokes java lib. java11:
Exception in thread "main" java.lang.NoClassDefFoundError: org/gradle/internal/classpath/Instrumented
at org.apache.velocity.util.ExtProperties$1.run(ExtProperties.java:187)
at java.base/java.security.AccessController.doPrivileged(Native Method)
at org.apache.velocity.util.ExtProperties.<init>(ExtProperties.java:182)
at org.apache.velocity.runtime.RuntimeInstance.<init>(RuntimeInstance.java:170)
at org.apache.velocity.app.VelocityEngine.<init>(VelocityEngine.java:55)
at net.masterthought.cucumber.generators.AbstractPage.<init>(AbstractPage.java:39)
at net.masterthought.cucumber.generators.ErrorPage.<init>(ErrorPage.java:18)
at net.masterthought.cucumber.ReportBuilder.generateErrorPage(ReportBuilder.java:229)
at net.masterthought.cucumber.ReportBuilder.generateReports(ReportBuilder.java:109)
at org.foo.bar.gradle.plugin.ReportGenerator.generateCompleteReport(ReportGenerator.java:42)
Caused by: java.lang.ClassNotFoundException: org.gradle.internal.classpath.Instrumented
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)

**Comment by e1m7bo** [comment]
I've following exception on 6.7:
java.lang.NoClassDefFoundError: org/gradle/internal/classpath/Instrumented
        at org.codehaus.groovy.reflection.GroovyClassValueFactory.<clinit>(GroovyClassValueFactory.java:32)
        at org.codehaus.groovy.reflection.ClassInfo.<clinit>(ClassInfo.java:109)
        at org.codehaus.groovy.reflection.ReflectionCache.getCachedClass(ReflectionCache.java:110)
        at org.codehaus.groovy.reflection.ReflectionCache.<clinit>(ReflectionCache.java:55)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.registerMethods(MetaClassRegistryImpl.java:214)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:112)
        at org.codehaus.groovy.runtime.metaclass.MetaClassRegistryImpl.<init>(MetaClassRegistryImpl.java:90)
        at groovy.lang.GroovySystem.<clinit>(GroovySystem.java:36)
        at org.codehaus.groovy.runtime.InvokerHelper.<clinit>(InvokerHelper.java:82)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.createMap(ScriptBytecodeAdapter.java:636)

**Comment by e1m7bo** [comment]
Is there a chance to have it fixed in of next release?

**Comment by racheruk** [comment]
This one is a blocker for us. :disappointed: 

Is there a work around for this issue ? 

We are currently using gradle 6.5.1 and running into https://github.com/gradle/gradle/issues/14168 while using latest protobuf. That issue fix is part of 6.7 and we can not upgrade because of this issue. 

For us the exception is thrown while one of the custom plugin trying to use ClassLoader.getSystemClassLoader()

Exception for reference:
```
Exception in thread "main" java.lang.NoClassDefFoundError: org/gradle/internal/classpath/Instrumented
        at com.abcd.gradle.plugins.SomePlugin.$instrumentedCallSiteArray(SomePlugin.groovy)
        at com.abcd.gradle.plugins.SomePlugin.$getCallSiteArray(SomePlugin.groovy)
        at com.abcd.gradle.plugins.SomePlugin.<clinit>(SomePlugin.groovy:15)
        at com.abcd.SomeClass.finalizeBody(SomeClass.java:262)
        at com.abcd.SomeClass.getBody(SomeClass.java:184)
        at com.abcd.SomeClass.generateMultiplePageHtml(SomeClass.java:225)
        at com.abcd.SomeClass.generateHtmlFromMarkdownSources(SomeClass.java:132)
        at com.abcd.SomeClass.generate(SomeClass.java:91)
        at com.abcd.SomeClass.main(SomeClass.java:79)
Caused by: java.lang.ClassNotFoundException: org.gradle.internal.classpath.Instrumented
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        ... 9 more
```

**Comment by donat** [comment]
I've also encountered this when updating the build scripts in Buildship to 6.8-rc-1. You can reproduce it with [my feature branch](https://github.com/eclipse/buildship/tree/donat/update-build-scripts): just execute `./gradlew installTargetPlatform`. 

The stacktrace points to `org.apache.maven.artifact.ant.AbstractArtifactTask.initSettings()`.
```
Caused by: java.lang.NoClassDefFoundError: org/gradle/internal/classpath/Instrumented
        at org.apache.maven.artifact.ant.AbstractArtifactTask.initSettings(AbstractArtifactTask.java:237)
        at org.apache.maven.artifact.ant.AbstractArtifactTask.execute(AbstractArtifactTask.java:750)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
        ... 149 more
Caused by: java.lang.ClassNotFoundException: org.gradle.internal.classpath.Instrumented
        at org.apache.tools.ant.AntClassLoader.findClassInComponents(AntClassLoader.java:1383)
        at org.apache.tools.ant.AntClassLoader.findClass(AntClassLoader.java:1338)
        at org.apache.tools.ant.AntClassLoader.loadClass(AntClassLoader.java:1093)
        ... 156 more
``` 

`AbstractArtifactTask.java:237` does try to call `System.getProperty("user.home")`.

**Comment by Vampire** [comment]
Here another instance that is probably related:
https://github.com/ph4r05/javacard-gradle-plugin/issues/7
There the Gradle plugin uses a dirty code source location lookup trick to get a JAR from the current class path and then adds this jar to the class path of a `JavaExec` task (sub class) that then fails at runtime due to the missing instrumentation class.

I cannot see from the posted examples whether the other cases are the same, but I guess that it is at least similar, that the plugins use the build script class path for other things and then the instrumentation fails.

In my case it was easy to solve, I just removed the bad jar from those class paths, as the jar was added to it in a different way anyway, so the lookup and adding was totally unnecessary.

Maybe the instrumentation could be enhanced in a way so that it works within Gradle context and without, so for example checking whether the class `Instrumented` is present, if yes, then use the instrumented code, if not then use the original code.

**Comment by e1m7bo** [comment]
> There the Gradle plugin uses a dirty code source location lookup trick to get a JAR from the current class path and then adds
> this jar to the class path of a JavaExec task (sub class) that then fails at runtime due to the missing instrumentation class.

@Vampire It seems that I have the same problem but with this exception that I cannot remove those additional jars from classpath (i.e. my Plugin is running JavaExec to scan classes/resources available on build classpath). So unfortunately I cannot use your workaround. 

I'm blocked with old version of Gradle now for over half a year :(

**Comment by Vampire** [comment]
Well, another work-around would probably be to either add the jar with the needed class to the class path too, or have a dummy implementation of the instrumentation class.

**Comment by racheruk** [comment]
It would be good to know the timeline when such a blocker/critical issue will be addressed. Without that it feels a bit constrained with sticking to old release.

**Comment by trustin** [comment]
Can we disable instrumentation or make it optional for certain dependencies? I think the solution suggested by @Vampire will work as well.

**Comment by kyay10** [comment]
Can confirm that this bug is still there on version 6.7.1

**Comment by racheruk** [comment]
Can confirm this issue persists in 6.8.x as well.

**Comment by e1m7bo** [comment]
I've just tested on 6.8.3 - problem still occurs.

**Comment by rachelcarmena** [comment]
**Update**: it also happens with latest `7.0`.

**Comment by racheruk** [comment]
Hope someone from gradle team listening :pray: 
Sincerely request for any thoughts on when this can be fixed. Thanks in advance.

**Comment by ysb33r** [comment]
Reported as a bug on Ivypot plugin as well https://gitlab.com/ysb33rOrg/ivypot-gradle-plugin/-/issues/49

**Comment by olegbolden** [comment]
> The error comes from the custom Gradle Plugin that provides the Kotlin Compiler Plugin.
> 
> No errors when providing the Kotlin Compiler Plugin manually via `compileKotlin.kotlinOptions.freeCompilerArgs` without the custom Gradle Plugin.

rachelcarmena,
Can you kindly show configuration example for this opiton?

**Comment by rachelcarmena** [comment]
> Can you kindly show configuration example for this opiton?

Of course:

* Where the Gradle Plugin provides the Kotlin Compiler Plugin: https://github.com/arrow-kt/arrow-meta/blob/main/gradle-plugin-commons/src/main/kotlin/arrow/meta/plugin/gradle/ArrowMetaGradlePlugin.kt#L30

* When providing the Kotlin Compiler Plugin manually via `compileKotlin.kotlinOptions.freeCompilerArgs` without the custom Gradle Plugin: https://github.com/arrow-kt/arrow-meta-examples/blob/main/generate-sources/use-plugin/build.gradle#L28

However, the use of the Kotlin Compiler Plugin is not relevant for this issue. I mean, I would like to say that the error came from the **Gradle Plugin**:

* Adding some configuration things **directly** in a Gradle file: :heavy_check_mark: 
* Embedding that configuration in a **Gradle Plugin**: :red_circle: 

Fortunately, we could remove the use of `System.getProperties()` to progress (they weren't essential to make the plugin work but to show debug info).

I hope we can recover that additional feature some day :pray:

**Comment by olegbolden** [comment]
Thank you very much!
Hope, this will be helpful to someone else.

**Comment by NicolasRouquette** [comment]
We have experienced this problem with custom gradle tasks in this project: https://github.com/opencaesar/owl-tools

The only recourse we have at this time is to stick to Gradle 6.5.1

**Comment by rachelcarmena** [comment]
Thanks for sharing your issue @NicolasRouquette :raised_hands:  I think it's the first reported case here about a Gradle Task Type.

It seems something related to classes management changed from Gradle 6.5.1 that makes the use of custom Gradle Plugins or custom Gradle Task Types fail when calling `ClassLoader.getSystemClassLoader()` (your case as well as other cases reported in this thread) or `System.getProperties()`.

I hope these clues are useful to the Gradle Team :pray:

**Comment by Vampire** [comment]
Why clues? ANd why first reproted case?
It is the exact same problem.
It is exactly clear why it is happening.
And I also reported this happened in a 3rd party task.
That plugin - at least that's how @NicolasRouquette described it - looks in its own classpath for a certain jar and then uses it outside the build script runtime classpath and that simply is not supported currently.

That plugin could easily do it properly and so make itself compatible with newer Gradle versions just like the plugin I mentioned did.
But noone opened an issue in their tracker yet to tell them that their plugin is not compatible.

**Comment by racheruk** [comment]
Frankly, i am not even sure why there is no workaround provided so far. Let alone fixing it. Bit disappointing actually :(

**Comment by NicolasRouquette** [comment]
> That plugin - at least that's how @NicolasRouquette described it - looks in its own classpath for a certain jar and then uses it outside the build script runtime classpath and that simply is not supported currently.

Is this limitation documented somewhere?

> That plugin could easily do it properly and so make itself compatible with newer Gradle versions just like the plugin I mentioned did.

@Vampire Can you recommend either a documentation or an example for a proper class lookup technique?

**Comment by Vampire** [comment]
> Frankly, i am not even sure why there is no workaround provided so far. Let alone fixing it. Bit disappointing actually :(

Well, there is a "work-around" or actually fix, namely don't try to use jars from the build script class path outside the build script class path, but instead use detached configurations to declare and resolve the libraries you need, then you also automatically get the transitive dependencies. It is just bad practice to search for a jar in the current class path and then build a new class loader from it. This also is a bit fragile as the class loader could anytime be changed and not use files on disk to load classes from but dynamically generate them, network-load them or whatever. If you get struck by this issue, then the part that causes it like the `owl-tools` plugin is following bad practice which just now stopped working as there were internal changes.

While it could probably be made in Gradle so that it would still work, the actual bug is in the unclean plugins using jars from the build script class path and there the bugs should be fixed, even if this concrete breakage gets worked-around within Gradle code as there could as well be future changes breaking this again.

> Is this limitation documented somewhere?

I don't know of any documentation, but as I just described in this comment again, it is bad practice that works with specific Gradle versions as it uses internal knowledge on how the class loading is currently (or better was up to now) implemented and thus uses tight-coupling which should be avoided. Those plugins are simply not compatible with newer Gradle versions (unless this gets worked-around within Gradle) due to this tight coupling. The point is not, that this limitation is not documented. The point is, that the plugins doing such actions rely on internal behaviour that is not documented and not part of the API and thus can change anytime and thus break. which is why tight-coupling is most often a bad idea.

> @Vampire Can you recommend either a documentation or an example for a proper class lookup technique?

I just described it earlier in this comment one way and you could also look at the commit where the plugin I mentioned here earlier fixed it in a different way to become compatible with newer Gradle versions again.

Basically you can either create a detached configuration, declare your dependencies on it and resolve it, or the more flexible and user-friendly approach that works even if coordinates change and similar is to create a normal configuarion, and add the needed libs as default dependenices, then the user of the plugin has the possibility to declare different dependencies even if the coordinates changed.

**Comment by aburaksahin** [comment]
@ljacomet @eskatos Could you guys give us an update on this issue? It really hampers our workflow.

**Comment by ljacomet** [comment]
Going over this issue, I am afraid the answer will be to not try to load _transformed_ buildscript dependencies outside of a Gradle execution context.
Instead you should use proper dependency management to pull in what you need.

@eskatos Anything to add?

**Comment by OmriYaHoo** [comment]
Hi @Vampire,
Could you please elaborate on this?

> Basically you can either create a detached configuration, declare your dependencies on it and resolve it, or the more flexible and user-friendly approach that works even if coordinates change and similar is to create a normal configuration, and add the needed libs as default dependencies, then the user of the plugin has the possibility to declare different dependencies even if the coordinates changed.

We have our own custom plugin for internal usage by the teams. We have the following code where the line with the `System.getProperty` is throwing the exception mentioned in the issue.  
```java
private static List<URL> getAllClassPath(ClassLoader loader) {
    if (loader instanceof URLClassLoader) {
        return ImmutableList.
                <URL>builder().
                add(((URLClassLoader) loader).getURLs()).
                addAll(getAllClassPath(loader.getParent())).
                build();
    } else {
        return ImmutableList.
                copyOf(System.getProperty("java.class.path").split(":")).
                stream().
                map(ClassPathHandler::getUrl).
                collect(Collectors.toList());
    }
}
```
This method is used initially with the class loader of the class that it's part of, meaning it's called like this -  `getAllClassPath(ClassPathHandler.class.getClassLoader()`.
I wasn't the one who wrote this code so I'm struggling a bit with how to resolve this issue.  

I tried to understand the solution for the `ivypot-gradle-plugin` you shared but couldn't get a good grasp on what I'm supposed to implement to fix this issue in our case.
As far as I understood the `Instrumented` class is taking control of the `System.getProperty` [part](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/internal/classpath/AccessTrackingProperties.java#L391) (This was extracted to the `AccessTrackingProperties` class instead of the `Instrumented`) but this class exists from both the Plugin classpath we have, and also from the service classpath we are using this plugin since both of them are using Gradle 6.9.1 which includes the `org.gradle.internal.classpath.Instrumented` class. Why would this error happen?  
I'm assuming that this class is also included in the Plugin when the service calls its Tasks.
**Could you please share a code example of how to implement this detached configuration or any other of the solutions you suggested?** I prefer not to implement a dirty workaround.  

Regrading this:
> Going over this issue, I am afraid the answer will be to not try to load transformed buildscript dependencies outside of a Gradle execution context.
Instead you should use proper dependency management to pull in what you need.

What does this mean exactly? We are already using the Gradle dependency management. Are we supposed to import the plugin code as a dependency instead of a Plugin declaration?

Sorry for the number of questions I throw at you, this thread has many abstract definitions and it's a bit hard to understand the whole picture like you the maintainers. I try to read the Gradle docs/blogs but couldn't find any useful information on how to solve this issue.

**Comment by OmriYaHoo** [comment]
@ljacomet @eskatos could any of the Gradle team address the above message I posted?

**Comment by NicolasRouquette** [comment]
FYI: I managed to work around this problem inspired by @Vampire's comments about using a gradle configuration classpath. This helps me migrate from Gradle 6.5.1 to 7.4.2.

Here is the gist of the solution in the `build.gradle` script:

```
plugins {
    id 'com.github.psxpaul.execfork' version '0.2.0'
}

configurations {
    fuseki
}

dependencies {
    fuseki 'org.apache.jena:jena-fuseki-server:4.2.0'
}

/*
 * Start the headless Fuseki server as a background build process.
 * Note that the server is stopped after the build; this alleviates the need for a stopFuseki task.
 */
task startFuseki(type: com.github.psxpaul.task.JavaExecFork, group:"oml") {
    classpath = project.files().from(configurations.fuseki.resolve().toList())
    main = 'org.apache.jena.fuseki.main.cmds.FusekiMainCmd'
    def config = file('.fuseki.ttl')
    args = [ "--config=$config" ]
    waitForPort = 3030
    workingDir = file('.fuseki')
    standardOutput = file('.fuseki/fuseki.log')
}
```

Hope this helps!

- Nicolas.

**Comment by mlopatkin** [comment]
So, to put a summary here:
1. The JAR files on the build script classpath are not the original JARs because Gradle modifies the bytecode of plugins and their dependencies to do various internal tasks. In particular, some Java API calls are replaced with calls to Gradle's Instrumented class to intercept them and detect build script inputs like system properties and environment variables.
2. Because of (1) it is not possible to use JARs of the build script classpath outside the original Gradle's classloader, at least not without the significant effort. It may work for now if the code doesn't hit instrumented call sites, but it may break in the future versions of Gradle when the list of instrumented API calls grows.
3. Therefore, constructing a classpath out of JARs retrieved from the build script's classpath (with `protectionDomain.codeSource.location` or casting the classloader to `URLClassLoader`) is not supported by Gradle and should be avoided.
4. As a preferred alternative, one could use Gradle's dependency management capabilities to obtain necessary JARs, e.g. by declaring a configuration with these JARs as dependencies and resolving it. The Worker API with classpath/process isolation could also be of help there.

Eventually, we might change the way instrumentation works to avoid exposing the modified JARs at all, but I cannot provide any estimates of when it happens. We'd still prefer users to switch to supported ways of getting dependencies.
